src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java

```diff
@@ -400,26 +406,40 @@ public final void onEvent(@Nonnull GenericEvent event)
         else if (event instanceof ReconnectedEvent)
             onReconnect((ReconnectedEvent) event);
 
-        Class<ListenerAdapter> handle = ListenerAdapter.class;
         for (Class<?> clazz : ClassWalker.range(event.getClass(), GenericEvent.class))
         {
-            String name = clazz.getSimpleName();
-            try
-            {
-                name = "on" + name.substring(0, name.length() - "Event".length());
-                Method method = handle.getDeclaredMethod(name, clazz);
-                method.invoke(this, event);
-            }
-            catch (NoSuchMethodException | IllegalAccessException ignored) {} // this means this is probably a custom event!
-            catch (InvocationTargetException e)
-            {
-                Throwable cause = e.getCause();
-                if (cause instanceof RuntimeException)
-                    throw (RuntimeException) cause;
-                if (cause instanceof Error)
-                    throw (Error) cause;
-                throw new RuntimeException(cause);
-            }
+            MethodHandle mh = methods.computeIfAbsent(clazz, ListenerAdapter::findMethod);
```

Comment:
ConcurrentHashMap won't enter the value if it is null
> If the specified key is not already associated with a value, attempts to compute its value using the given mapping function and **enters it into this map unless null**.

which means it won't cache the result when it can't find the method (most notably that happened for Event and Object) removing a lot of the optimization (from my benchmark this actually ended up being over 4x slower than the current method)